solutionbox/code_free_ml/mltoolbox/code_free_ml/data/trainer/task.py

```diff
@@ -0,0 +1,849 @@
+  elif args.model_type == 'linear_classification':
+    estimator = tf.contrib.learn.LinearClassifier(
+        feature_columns=feature_columns,
+        n_classes=target_vocab_size,
+        config=config,
+        model_dir=train_dir,
+        optimizer=tf.train.AdamOptimizer(
+            args.learning_rate, epsilon=args.epsilon))
+  else:
+    raise ValueError('bad --model-type value')
+
+  return estimator
+
+
+def read_vocab(args, column_name):
+  """Reads a vocab file if it exists.
+
+  Args:
+    args: command line flags
+    column_name: name of column to that has a vocab file.
+
+  Returns:
+    List of vocab words or [] if the vocab file is not found.
+  """
+  vocab_path = os.path.join(args.analysis_output_dir, VOCAB_ANALYSIS_FILE % column_name)
+
+  if not file_io.file_exists(vocab_path):
+    return []
+
+  vocab_str = file_io.read_file_to_string(vocab_path)
+  vocab = pd.read_csv(six.StringIO(vocab_str),
+                      header=None,
+                      names=['token', 'count'],
+                      dtype=str)  # Prevent pd from converting numerical categories.)
+  return vocab['token'].tolist()
+
+
+def get_target_name(features):
+  for name, transform in six.iteritems(features):
+    if transform['transform'] == TARGET_TRANSFORM:
+      return name
+  return None
+
+
+def get_key_name(features):
+  for name, transform in six.iteritems(features):
+    if transform['transform'] == KEY_TRANSFORM:
+      return name
+  return None
+
+
+def gzip_reader_fn():
+  return tf.TFRecordReader(options=tf.python_io.TFRecordOptions(
+      compression_type=tf.python_io.TFRecordCompressionType.GZIP))
+
+
+def get_experiment_fn(args):
+  """Builds the experiment function for learn_runner.run.
+
+  Args:
+    args: the command line args
+
+  Returns:
+    A function that returns a tf.learn experiment object.
+  """
+
+  def get_experiment(output_dir):
+    # Merge schema, input features, and transforms.
+    schema_file_path = os.path.join(args.analysis_output_dir, SCHEMA_FILE)
+    if not file_io.file_exists(schema_file_path):
+      raise ValueError('File not found: %s' % schema_file_path)
+    schema = json.loads(file_io.read_file_to_string(schema_file_path).decode())
+
+    features_file_path = os.path.join(args.analysis_output_dir, FEATURES_FILE)
+    if not file_io.file_exists(features_file_path):
+      raise ValueError('File not found: %s' % features_file_path)
+    features = json.loads(file_io.read_file_to_string(features_file_path).decode())
+
+    stats_file_path = os.path.join(args.analysis_output_dir, STATS_FILE)
+    if not file_io.file_exists(stats_file_path):
+      raise ValueError('File not found: %s' % stats_file_path)
+    stats = json.loads(file_io.read_file_to_string(stats_file_path).decode())
+
+    target_column_name = get_target_name(features)
+    key_column_name = get_key_name(features)
+    header_names = [col['name'] for col in schema]
+    if not target_column_name or not key_column_name:
```

Comment:
This is for future work item: I think key should be optional (match prediction API).